newm-server/src/main/kotlin/io/newm/server/features/cardano/parser/NFTSongParser.kt

```diff
@@ -0,0 +1,142 @@
+package io.newm.server.features.cardano.parser
+
+import io.ktor.utils.io.core.toByteArray
+import io.newm.chain.grpc.LedgerAssetMetadataItem
+import io.newm.chain.grpc.NativeAsset
+import io.newm.server.features.cardano.model.NFTSong
+import io.newm.shared.koin.inject
+import io.newm.shared.ktx.debug
+import io.newm.shared.ktx.warn
+import org.koin.core.parameter.parametersOf
+import org.slf4j.Logger
+import java.util.UUID
+import kotlin.time.Duration
+
+private val logger: Logger by inject { parametersOf("NFTSongParser") }
+
+fun List<LedgerAssetMetadataItem>.toNFTSongs(asset: NativeAsset): List<NFTSong> {
+    logger.debug { "Parsing PolicyID: ${asset.policy}, assetName: ${asset.name}" }
+
+    var image: String? = null
+    var songTitle: String? = null
+    var songDuration: String? = null
+    val artists = mutableSetOf<String>()
+    val genres = mutableSetOf<String>()
+    val moods = mutableSetOf<String>()
+    val files = mutableSetOf<File>()
+
+    val parseSongItem = { item: LedgerAssetMetadataItem ->
+        when (item.key) {
+            "artists" -> {
+                artists += item.artists()
+            }
+
+            "genres" -> {
+                genres += item.values()
+            }
+
+            "mood" -> {
+                moods += item.value
+            }
+
+            "song_title" -> {
+                songTitle = item.value
+            }
+
+            "song_duration" -> {
+                songDuration = item.value
+            }
+        }
+    }
+
+    for (item in this) {
+        when (item.key) {
+            "image" -> {
+                image = item.value
+            }
+
+            "files" -> {
+                files += item.audioFiles()
+            }
+
+            "release" -> {
+                // only available in CIP60-V2 or above
+                item.childrenList.forEach(parseSongItem)
+            }
+
+            else -> parseSongItem(item)
+        }
+    }
+    return files.mapNotNull { file ->
+        val title = file.songTitle ?: songTitle ?: file.name
+        if (title == null || file.src == null || image == null) {
+            logger.warn { "Skipped invalid PolicyID: ${asset.policy}, assetName: ${asset.name}, title = $title, src: ${file.src}, image = $image" }
+            null
+        } else {
+            NFTSong(
+                id = UUID.nameUUIDFromBytes(file.src.toByteArray()),
+                policyId = asset.policy,
+                assetName = asset.name,
+                amount = asset.amount.toLong(),
+                title = title,
+                audioUrl = file.src.toResourceUrl(),
+                imageUrl = image.toResourceUrl(),
+                duration = (file.songDuration ?: songDuration)?.let { Duration.parse(it).inWholeSeconds } ?: 0,
```

Comment:
Should we use -1 instead of 0 for an unknown duration?

Reply:
I agree. Done